### ISSUE TITLE
Issues using additional fonts in user fonts folders (SVG/Loading/Rendering)

### ISSUE THREAD

**Issue by Phillipus** [archimatetool/archi#628]
Windows 10
Archi 4.6

1. Install an additional font in Windows
1. On my system this is actually installed in `C:\Users\UserName\AppData\Local\Microsoft\Windows\Fonts`
1. Use this font in a View
1. Export image to either SVG (choose "Embed fonts" option) or PDF
1. The font is not used

However, if I use a font that is present in `C:\Windows\Fonts` it does get exported.

I wonder if this is some kind of loading issue?

Windows will not actually install a user font in `C:\Windows\Fonts` but in `C:\Users\UserName\AppData\Local\Microsoft\Windows\Fonts`

We need to test if we can actually install a font into `C:\Windows\Fonts` and see if it works.

**Comment by jbsarrodie** [comment]
FWIW, I never got "Embed fonts" to work on Linux with SVG (did not test PDF). There might be a common root case.

**Comment by Phillipus** [comment]
SVG and PDF use the same code pretty much:

1. Render the View to `ExtendedGraphicsToGraphics2DAdaptor` onto a `SVGGraphics2D`
2. Then the PDF takes that SVG and transcodes it

**Comment by Phillipus** [comment]
Now this is very interesting...

If you install a font on Windows by right-clicking the .ttf file and choose "Install for all users" then the font is installed in `C:\Windows\Fonts` and it works OK when exporting to SVG/PDF.

Now why is the SVG library not using fonts that are not installed for all users?

**Comment by Phillipus** [comment]
If you open up an exported SVG file that should have a font in it (not installed for all users) there's no reference to it at all. It's as if one of the SVG libraries is not even loading it.

**Comment by Phillipus** [comment]
Intuition tells me something to do with `java.awt` loading fonts in a different way and using the `GraphicsToGraphics2DAdaptor`...

**Comment by Phillipus** [comment]
For reference - https://issues.apache.org/jira/browse/BATIK-872

**Comment by Phillipus** [comment]
@jbsarrodie This will affect the Font Loader plug-in. Fonts in user directory will not export to SVG,

**Comment by Phillipus** [comment]
> Intuition tells me something to do with java.awt loading fonts in a different way

I was right. It seems that somewhere in either `SVGGeneratorContext`, `SVGGraphics2D` or`GraphicsToGraphics2DAdaptor` java.awt is only loading fonts found in `C:\Windows\Fonts`

If we hack the export code with this:

```
GraphicsEnvironment lge = GraphicsEnvironment.getLocalGraphicsEnvironment();
for(File fontFile : new File("C:\\Users\\Phillipus\\AppData\\Local\\Microsoft\\Windows\\Fonts").listFiles()) {
    try {
        Font font = Font.createFont(Font.TRUETYPE_FONT, fontFile);
        lge.registerFont(font);
    }
    catch (Exception ex) {
    }
}
```

then the user fonts are loaded and it works.

So now we have to investigate where the fault lies...is it `java.awt`?

**Comment by Phillipus** [comment]
Comment here - https://stackoverflow.com/questions/22681771/java-cant-see-all-installed-fonts-in-system

> On some Windows machines there are two buttons for installing fonts: Install For Me and Install For All Users. Java only lists fonts installed for all users.

**Comment by Phillipus** [comment]
I'm not finding any info on this other than manually loading fonts as per my hack above.

We could leverage the Font Loader plug-in to load user fonts into `GraphicsEnvironment` as a workaround for this. Otherwise, document it as a known issue that user fonts have to be installed for all users for SVG/PDF export.

I tested on Mac and I don't think this affects it. Not sire about Linux.

**Comment by Phillipus** [comment]
> We could leverage the Font Loader plug-in to load user fonts into `GraphicsEnvironment` as a workaround for this,

@jbsarrodie I've created a branch in Font Loader plug-in, `awt-font` that does exactly this. I need to research this some more, though. Ideally we would like the Batik library to load all fonts on Windows.

**Comment by Phillipus** [comment]
Loading the fonts into AWT only works if the font has been installed by the user (and exists in the `C:\Users\UserName\AppData\Local\Microsoft\Windows\Fonts folder`)

So loading a font from a known file location into `GraphicsEnvironment` won't work.

**Comment by Phillipus** [comment]
Summary of findings:

- **Windows**: the font has to be installed for all users or we can manually load the font into AWT if the font has been installed (not for all users) and exists in the `C:\Users\UserName\AppData\Local\Microsoft\Windows\Fonts` folder
- **Mac**: it just works
- **Linux**: no embedded fonts work at all

**Comment by Phillipus** [comment]
For reference - https://stackoverflow.com/questions/48152492/apache-batik-1-7-apache-fop-1-1-not-rendering-fonts-on-ubuntu

Note having to use a config file

**Comment by Phillipus** [comment]
#552 might be related

**Comment by Phillipus** [comment]
This might be useful regarding font sizes:

https://bugs.eclipse.org/bugs/show_bug.cgi?id=312723

They have a hack for `GraphicsToGraphics2DAdaptor`:

https://git.eclipse.org/c/gmf-runtime/org.eclipse.gmf-runtime.git/commit/?id=2d71b27e2d536c459ae115f92f2b9a5790454078

**Comment by Phillipus** [comment]
There's a new branch "awt-fonts". This loads user fonts in the "fonts" folder into AWT.

But something is strange.

1. Install a font into Archi's "fonts" folder
2. Create a simple View with one object in it and use that font
3. Export to SVG image and ensure that "embed fonts" is ticked

Now open the SVG file in a text editor. You'll see that the font is embedded and each text character is rendered in SVG as a glyph.

But...

if I open the file in Edge, Firefox, or Chrome the font does not display. But on Mac Safari it does.

If I create a simple View using two objects with two fonts - one font is installed for all users and one font is installed in the "fonts" folder then the first font displays but not the second. On Mac Safari both fonts display OK.

[Default View.svg.zip](https://github.com/archimatetool/archi/files/4702509/Default.View.svg.zip)

**Comment by Phillipus** [comment]
In the attached example the first font in the top box is a user font and does not display, the second font is installed on the system and does display.

Open the SVG file in a text editor - both fonts are embedded!

Why?

[Default View.svg.zip](https://github.com/archimatetool/archi/files/4702544/Default.View.svg.zip)

**Comment by Phillipus** [comment]
In this attached example the first font is now installed on the system. Compare the SVG in a text file.

The first thing I notice is the precision of decimal points:

With the font installed in Archi's font folder the SVG font XML looks like this:

```
<font-face ascent="79.59992"
<glyph unicode="n" horiz-adv-x="47.89995"
<glyph unicode="i" horiz-adv-x="24.899979"
```

With the font installed in the system the SVG font XML looks like this:

```
<font-face ascent="79.6"
<glyph unicode="n" horiz-adv-x="47.9"
<glyph unicode="i" horiz-adv-x="24.9"
```

[Default View Installed.svg.zip](https://github.com/archimatetool/archi/files/4702606/Default.View.Installed.svg.zip)

**Comment by Phillipus** [comment]
Cancel that last comment. It has nothing to do with anything.

It seems that Safari will display the SVG properly.

**Comment by Phillipus** [comment]
Re-opening. That close issue button is too big.

**Comment by Phillipus** [comment]
Export to PDF is working OK with this fix. That's because it is first converted to SVG then converted to a PDF file and is self-contained.

It seems that some browsers won't display the fonts in an SVG file unless the font is installed on the system even if the font info is embedded.

**Comment by Phillipus** [comment]
The AWT font loading patch works for Windows and Mac but not for Linux.

On Linux `GraphicsEnvironment#registerFont()` returns false.

On Windows and Mac, user fonts _are_ embedded in the SVG file but you need the right kind of SVG viewer to see them. So far I could only find Mac Safari that will display them correctly.

On Window and Mac, user fonts _are_ embedded in the PDF file and will display correctly in all PDF viewers.

So it seems that we have several problems:

- Linux won't register fonts into AWT therefore they won't get embedded in SVG/PDF
- Windows/Mac will embed fonts into AWT but browsers and other SVG viewers won't display the font unless it is installed on the system (with the exception of Safari for Mac)

So far we're good for exporting to PDF on Windows and Mac with this patch.

**Comment by Phillipus** [comment]
Actually it turns out the AWT font loading patch has an effect only on Windows and this can be demonstrated when exporting the image to PDF. On Mac, it is not needed, and on Linux fonts are not loaded.

**Comment by Phillipus** [comment]
This talks about embedding fonts only on the system.

https://bz.apache.org/bugzilla/show_bug.cgi?id=46374

**Comment by jbsarrodie** [comment]
> if I open the file in Edge, Firefox, or Chrome the font does not display. But on Mac Safari it does.

Well, here is the answer: https://caniuse.com/#feat=svg-fonts

In short, SVG fonts never really got adopted and support has been dropped from most browsers. FWIW, even inkscape seems to no more support them (I've just tried without success).

The only known workaround is to use a CSS font-face definition. See https://oreillymedia.github.io/Using_SVG/extras/ch07-dataURI-fonts.html

That's what I tested some days ago with some success. This worked in browser but not in Inkscape, maybe because I not only use a custom font, but also ligature. See attached example:

[Default View.svg.zip](https://github.com/archimatetool/archi/files/4703808/Default.View.svg.zip)

**Comment by Phillipus** [comment]
> Well, here is the answer: https://caniuse.com/#feat=svg-fonts

Ah, OK.

> The only known workaround is to use a CSS font-face definition

I wonder if Batik supports this yet?

**Comment by jbsarrodie** [comment]
> I wonder if Batik supports this yet?

I don't know, but that's not to hard to implement by ourselves if needed: this basically involves getting the list of fonts, then finding their files on the system, then encoding them in base64, then writing the @font-face CSS section (that's what I did manually for my test).

**Comment by Phillipus** [comment]
Might be useful - https://stackoverflow.com/questions/11851821/how-i-use-a-custom-font-with-the-batik-svg-library

**Comment by Phillipus** [comment]
To insert the `style` tag. In `com.archimatetool.export.svg.SVGExportProvider`:

```java
Element styleElement = document.createElementNS("http://www.w3.org/2000/svg", "style");
styleElement.setAttribute("type", "text/css");
styleElement.setTextContent("fontFaceInfo");
root.insertBefore(styleElement, root.getFirstChild());
```

**Comment by Phillipus** [comment]
> this basically involves getting the list of fonts, then finding their files on the system, then encoding them in base64, then writing the @font-face CSS section (that's what I did manually for my test).

So:

1. _Getting the list of fonts._ This could be done by taking the root Figure and iterating through its children to get a list of used Font classes in the diagram.
1. _Finding their files on the system._  Don't know how to do this yet.
1. _Encoding them in base64_. Base64 class from file to string.
1. _Writing the @font-face CSS section_. See above, that's known.

**Comment by Phillipus** [comment]
I can solve this for Windows and Mac using a different approach. And the resulting SVG will display in all Browsers and SVG viewers.

Linux still has a problem with using fonts for SVG if they are in the user fonts folder. But this problem is to do with Batik.

Checkout the `awt-fonts` branch. Fonts are no longer embedded. Text is drawn as shapes, no font information is needed.

**Comment by Phillipus** [comment]
@jbsarrodie I've just discovered a more serious problem with the user fonts folder on Windows, related to not using `ScaledGraphics` to fix the text clipping when zoomed bug.

1. Copy a font into the user "fonts" folder
1. Use that font in a View
1. Zoom to > 100%.
1. The font is not drawn correctly and is obscured.

However, if you install the font into the system it is OK.

If we revert our changes involving not using `ScaledGraphics` (`ExtendedScalableFreeformRootEditPart` class) the font is rendered OK (apart from clipping at the end of text).

(This does not occur on Mac.)

**Comment by Phillipus** [comment]
To summarise :

- SVG/PDF export works on all platforms if we use the draw text as shapes options

Using fonts in the user fonts folder is causing two problems:

- Zoomed text is not working on Windows (Views and export > 100% to image)
- Fonts are not exported in SVG on Linux

The Zoomed text issue on Windows is a major problem.

If we don't have a user fonts folder all of the above problems are solved.

**Comment by kentxu** [comment]
My 2 cents  from past experience. 

The text distortion when stage is zoomed only happens on windows and only when the font is loaded through java (not installed by OS). 

This is a ScaledGraphics scale text vs native Graphic transform issue. 

With native Graphic scale, the java  loaded font renders text almost illegible when you zoom in/out. However, for font  installed on OS, the native way is superior as it scales text proportionally  in precision.   Too bad it seems there is no way to  workaround  the distortion  if you must use font loaded from java. OS simply does  not know how to scale that font properly.   I don't know if  that's something SWT team can do anything about. 

If font must be loaded through AWT in this app, then ScaledGraphics   is your best bet. At least if  you still want stage  zoom and would let user choose arbitrary   text  size.  As  you have mentioned, this is also what  draw2d/GEF's default implementation.  There are some  problems with draw2d's ScaledGraphics. 

1.   the  font size is calculated based on the scale value. This  causes  a natural accuracy loss. 10PT at 50% zoom is fine.  But it has  to be  rounded off  to  7PT if the  zoom level is  at  75%. This would explain why some  zoom value works better than the others.

2. The OS does not render text   at 50% when the font size is  halved. It could be longer or shorter based on the font  and the native OS algorithm.  It also  seems that java loaded font  renders slightly differently in  width than when the  same font is  installed  on the OS. It is  of  lesser quality (rigged edge) and  some of the stroke hints are  honored like how OS native does. I  don't think that's a surprise.   What that means, is that the  same text, at 5PT is not really 50% of that rendered at 10PT.  This  leads to various issues draw2d  text flow  system has  problem  with.

3. the  above  two problems caused the  text  to be cutoff at  times. In reality, it is sometimes longer  or shorter  than calculated (draw2d figureutils) .  Even more seriously, if there are multiple format in  the same line (richtext), the cutoff problems turns  into  a text  spacing issue. For example,  Draw2D TextFlow does  not take that into account as it assumes  text rendered in  5PT  would be  exactly half of  10PT.   


It  is possible to add some workaround/fix  to the  TextPage, InlineBlock and etc to fix most of the  problems,  except the scale inaccuracy as a whole  . Meaning, it is possible to make the text legible, without cutoffs and scaled close  enough. It is impossible to make it exact though  (of the scale value) due to #2  above. 

For the cutoff issue, the basic idea is to control clipping strategy of the  text page and  text box if needed depending  on your app's need. This has no impact to the container object's size, which  is very good.  It only  impacts  where and how the text can be overlaid by the  OS. 

It is  a bit  more complicated if  the  app supports rich  text formatting because each format (a  InlineBlock) within the same line  will cause some shift  in width for that  segment.  I am not super familiar with Archimate's situation, maybe this part is not relevant.

**Comment by jbsarrodie** [comment]
@kentxu Thank you for detailed explanation. This helps.

@Phillipus I still haven't had time to go back to code, but I think the best compromise is to remove fontloader feature from 4.7 and (try to) work on it for 4.8. I really prefer keeping current Archi feature set and removing zoom issues, than adding font loading feature and introducing new bugs.

Of course, this requires to fixe the issues you've noticed (related to borderwidth), but I think it's worth it.

FWIW, using an external tool such as WinFontsForAll on Windows does load fonts without having to install them, and Archi (ie the JVM) sees them as OS native. Sothere is a known workaround for users.

**Comment by Phillipus** [comment]
@kentxu Thanks very much for taking the time to explain things, we appreciate it.

@jbsarrodie I agree with that plan and we can try to work with not using `ScaledGraphics` adjusting the drawing code where possible (I've already done some of the figures). We can continue that discussion in #637 

I've spent some time looking at the source code for, and the issues around, `org.eclipse.draw2d.ScaledGraphics` and `org.eclipse.draw2d.SWTGraphics`.

The Draw2d framework is very old now (2002 or 2003) and I think in those days advanced drawing rendering was not always possible so they added the `ScaledGraphics` class. `ScaledGraphics` is basically a wrapper around `SWTGraphics` which, in turn, is a wrapper around `org.eclipse.swt.graphics.GC`.

At the lowest level it is the `GC` class that does the graphics rendering and makes calls to the OS. That's why there is a different `GC` class for each OS. `GC` is where the real drawing is done and any other classes like `ScaledGraphics` and `SWTGraphics` are just interfaces to it (handling things like clipping and font metrics).

So actually we only need to use the `SWTGraphics` class but there are some differences, especially in font handling. Draw2d figures (connection ends, for example) were written to use `ScaledGraphics` and that explains why we are seeing some artefacts now that we don't use it.

Ideally `SWTGraphics` should be revisited and adjusted to handle these edge cases on modern graphics. However that isn't going to happen as Draw2D is no longer maintained. I find this surprising as it the core framework underlying GEF, GMF, and the Sirius frameworks. I imagine that GMF and Sirius have their own hacks to deal with issues.

As I mentioned the Modelio modelling tool uses GEF and Draw2d and they have taken the decision to not use `ScaledGraphics` because of the font sizing issues.

There will always be a compromise when it comes to using the  `SWTGraphics` class (and Draw2d). It's a pity that the GEF developers didn't stick with drawing to SWT's `GC` class and instead went down the road of `JavaFX` for GEF4/5. I suspect that uptake of GEF4/5 will be minimal and will eventually go the way of the Dodo...

**Comment by Phillipus** [comment]
SVG font issues are solved for Archi 4.7 as fonts are no longer embedded. Also, on Windows, any fonts in `C:\Users\UserName\AppData\Local\Microsoft\Windows\Fonts` are loaded into AWT before exporting.
